### PR TITLE
Zero copy between Netty's `Http2Headers` and Armeria's `HttpHeaders`

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Netty's HTTP/2 codec decodes HTTP/2 headers to `Http2Headers` and
Armeria HTTP/2 codec converts `Http2Headers` into `RequestHeaders`,
`ResponseHeaders` and `HttpHeaders`.
It is inefficient since it copies headers two times.
After walking through Netty's API, I found out it is possible
to set a custom `Http2HeadersDecoder` that converts a headers block into
`HttpHeaders` when building `Http2ServerConnectionHandler`

Modifications:

- Add `ArmeriaHttp2Headers` that implements Netty's `Http2Headers` and
  internally stores the values in Armeria's `HttpHeaders`.
- Manually create `Http2ConnectionEncoder` and `Http2ConnectionDecoder`
  on both server and client side to customize `Http2HeadersDecoder`
- Update `ArmeriaHttpUtil` to directly convert `Http2Headers` to
  `HttpHeaders` without copies.

Result:

- Base work for HTTP/2 headers optimization.
- You can now trace HTTP/2 traffic by adjusting the logging level of `com.linecorp.armeria.logging.traffic.{server,client}.http2`.
  For example:  
  ```xml
  <!-- in your logback.xml -->
  <logger name="com.linecorp.armeria.logging.traffic.server.http2" level="TRACE" />
  <logger name="com.linecorp.armeria.logging.traffic.client.http2" level="TRACE" />
  ```

### CODE REVIEW

**Review by trustin — APPROVED** [review]
Looks like the good first step for optimizing HTTP/2 headers. Thanks for looking into this, @ikhoon !

File: core/src/main/java/com/linecorp/armeria/server/ServerHttp2ObjectEncoder.java [hunk 1]
```diff
@@ -82,22 +83,32 @@ public boolean isResponseHeadersSent(int id, int streamId) {
...
+            if (builder == null) {
+                builder = inputHeaders.toBuilder();
+            }
+            // We don't apply chunked encoding when the content-length header is set, which would
+            // prevent the trailers from being sent so we go ahead and remove content-length to force
+            // chunked encoding.
+            builder.remove(HttpHeaderNames.CONTENT_LENGTH);
+        }
+        if (builder == null) {
+            builder = inputHeaders.toBuilder();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/ServerHttp2ObjectEncoder.java:109]: How about assigning this builder at the first of the method and remove all `if (builder == null)`?

**Review by minwoox — COMMENTED** [review]
Left some minor comments.
Nice work!

File: core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java [hunk 2]
```diff
@@ -749,22 +731,21 @@ public static Http2Headers toNettyHttp2ServerHeaders(HttpHeaders inputHeaders) {
...
-            if (ADDITIONAL_RESPONSE_HEADER_DISALLOWED_LIST.contains(name)) {
-                continue;
-            }
-            if (isTrailerDisallowed(name)) {
-                continue;
-            }
-            outputHeaders.add(name, value);
+        final HttpHeadersBuilder builder = inputHeaders.toBuilder();
+
+        for (Entry<AsciiString, AsciiString> disallowed : HTTP_TO_HTTP2_HEADER_DISALLOWED_LIST) {
```
[the remark this answers is not in the document]
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java:736]: Now, I understand that we use this for case-insensitive comparison.
Let's change the name of the map as @trustin suggested. 😉

[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java:736]: `CaseInsensitiveMap` sounds good.

**Comment by anuraaga** [comment]
Any benchmark results? I remember trying an approach to zero copy a long time ago but found it to reduce performance and abandoned it. The approach was probably different than this one though don't remember - would be good to confirm the performance.

File: core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttp2Headers.java [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+    }
+
+    @Override
+    public boolean containsBoolean(CharSequence name, boolean value) {
+        return delegate.containsBoolean(name, value);
+    }
+
+    @Override
+    public boolean containsByte(CharSequence name, byte value) {
+        return getByte(name) == value;
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttp2Headers.java:364]: Good point!

File: core/src/main/java/com/linecorp/armeria/server/ServerHttp2ObjectEncoder.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/ServerHttp2ObjectEncoder.java:109]: Oops... I don't remember why I did like this.🤣
Maybe the code might be changed while fixing test cases.

**Comment by ikhoon** [comment]
I ran some benchmarks. Unfortunately, I didn't get any enhancement or regression. 😭
There is still room for performance improvement. I think the approach is not bad and could be a base work for further optimization.

**Review by minwoox — APPROVED** [review]
Thanks!

**Pull request merged**